Sources/Apollo/ConcatLink.swift

```diff
@@ -0,0 +1,41 @@
+extension NonTerminatingLink {
+  public func concat(_ next: NonTerminatingLink) -> NonTerminatingLink {
+    return NonTerminatingConcatLink(first: self, second: next)
+  }
+  
+  public func concat(_ next: TerminatingLink) -> TerminatingLink {
+    return TerminatingConcatLink(first: self, second: next)
+  }
+}
+
+fileprivate final class NonTerminatingConcatLink: NonTerminatingLink {
+  private var first: NonTerminatingLink
+  private var second: NonTerminatingLink
+  
+  init(first: NonTerminatingLink, second: NonTerminatingLink) {
+    self.first = first
+    self.second = second
+  }
+  
+  func request<Operation>(operation: Operation, context: LinkContext, forward: @escaping NextLink<Operation>) -> Promise<GraphQLResponse<Operation>> {
+    return first.request(operation: operation, context: context) {
+      return self.second.request(operation: $0, context: $1, forward: forward)
+    }
+  }
+}
+
+fileprivate final class TerminatingConcatLink: TerminatingLink {
+  private var first: NonTerminatingLink
+  private var second: TerminatingLink
+  
+  init(first: NonTerminatingLink, second: TerminatingLink) {
+    self.first = first
+    self.second = second
+  }
+  
+  func request<Operation>(operation: Operation, context: LinkContext) -> Promise<GraphQLResponse<Operation>> {
```

Comment:
We really can't rely on `Promise` here, because links need the ability to return multiple responses. That's important for subscriptions support, but also for something like a polling link. I'd rather not implement `Observable` (or rely on RxSwift), but maybe a response handler callback would work? On the other hand, promises make the link code a lot easier to write, and so would observables, so maybe it is worth implementing something minimal.

Reply:
Yeah, I know :(. 
Observables would be nice to have and it would probably be fun to implement, but difficult to get right.
Here's another idea: Similar to the cancel signal, we could add a `onNextResponse` callback to the context which is called whenever a link produces a new response. What do you think?

Reply:
Can you give an example of how `onNextResponse` would work? Would this replace returning a promise?